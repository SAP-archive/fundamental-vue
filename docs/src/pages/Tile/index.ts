import { ExampleCollectionFunction } from "../types";

export const plugin: ExampleCollectionFunction = () => {
  return {
    status: "inprogress",
    icon: "border",
    related: ["FdTile", "FdProductTile"]
  };
};
