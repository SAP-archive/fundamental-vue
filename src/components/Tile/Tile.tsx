import { Color, Colors, backgroundColorClassName } from '@/lib';
import { Slot, Component, Prop, Base } from '@/core';

interface Props {
    media?: any | null;
    title?: string | null;
    description?: string | null;
    actions?: any | null;
    isButton?: boolean;
    rowSpan?: number | null;
    colSpan?: number | null;
    backgroundColor?: Color | null;
}

@Component('Tile')
@Slot('media', 'Tile Media')
@Slot('actions', 'Tile Actions')
export class Tile extends Base<Props> {
    @Prop({ type: String, default: null })
    public title!: string | null;

    @Prop({ type: String, default: null })
    public description!: string | null;

    @Prop('row span', { type: Number, default: null })
    public rowSpan!: number | null;

    @Prop('column span', { type: Number, default: null })
    public colSpan!: number | null;

    @Prop('background color', { acceptableValues: Colors, default: null, type: String })
    public backgroundColor!: Color | null;

    @Prop('render tile as a button', { type: Boolean, default: false })
    public isButton!: boolean;

    @Prop('disable tile', { type: Boolean, default: false })
    public disabled!: boolean;

    public render() {
        const media = this.$slots.media;
        const actions = this.$slots.actions;
        const title = this.title;
        const description = this.description;
        const isButton = this.isButton;
        const disabled = this.disabled;

        return (
            <div class={this.classObject} role={isButton ? 'button' : null} aria-disabled={disabled}>
                {media && <div class='fd-tile__media'>{media}</div>}
                <div class='fd-tile__content'>
                    {title && <h2 class='fd-tile__title'>{title}</h2>}
                    {description && <p>{description}</p>}
                </div>
                {actions && <div class='fd-tile__actions'>{actions}</div>}
            </div>
        );
    }

    get classObject() {
        const rowSpanObject = () => {
            const rowSpan = this.rowSpan;
            if (typeof rowSpan === 'number') {
                return { [`fd-has-grid-row-span-${rowSpan}`]: true };
            }
            return {};
        };

        const colSpanObject = () => {
            const colSpan = this.colSpan;
            if (typeof colSpan === 'number') {
                return { [`fd-has-grid-column-span-${colSpan}`]: true };
            }
            return {};
        };

        const backgroundColorClasses =
            this.backgroundColor == null ? {} : { [backgroundColorClassName(this.backgroundColor)]: true };

        return {
            ...rowSpanObject(),
            ...colSpanObject(),
            ...backgroundColorClasses,
            'fd-tile': true,
        };
    }
}
