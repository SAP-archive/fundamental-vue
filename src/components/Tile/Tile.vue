<template>
  <div
    class="fd-tile"
    :class="classes"
    :role="role"
    :aria-disabled="disabled"
    v-fd-background-color="backgroundColor"
  >
    <div v-if="$slots.media" class="fd-tile__media">
      <slot name="media" />
    </div>
    <div class="fd-tile__content">
      <h2 v-if="title != null" class="fd-tile__title">{{ title }}</h2>
      <p v-if="description != null">{{ description }}</p>
    </div>
    <div v-if="$slots.actions" class="fd-tile__actions">
      <slot name="actions" />
    </div>
  </div>
</template>

<script>
import backgroundColor from "./../../directives/background-color";

export default {
  name: "FdTile",
  directives: { "fd-background-color": backgroundColor },
  props: {
    title: { type: String, default: null },
    description: { type: String, default: null },
    rowSpan: { type: Number, default: null },
    colSpan: { type: Number, default: null },
    backgroundColor: { type: String, default: null },
    isButton: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false }
  },
  computed: {
    role() {
      return this.isButton ? "button" : null;
    },
    classes() {
      return [
        this.rowSpan != null ? `fd-has-grid-row-span-${this.rowSpan}` : "",
        this.colSpan != null ? `fd-has-grid-column-span-${this.colSpan}` : ""
      ];
    }
  }
};
</script>
