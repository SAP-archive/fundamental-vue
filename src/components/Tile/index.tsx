export { Tile } from './Tile';
export { ProductTile } from './ProductTile';
