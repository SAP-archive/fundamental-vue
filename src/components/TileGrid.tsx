import { Component, Prop } from 'vue-property-decorator';
import { Api } from '@/api';
import { componentName } from '@/util';
import TsxComponent from '@/vue-tsx';

const colMapping = {
    2: '2-Column Grid',
    3: '3-Column Grid',
    4: '4-Column Grid',
    5: '5-Column Grid',
    6: '6-Column Grid',
};
type Col = keyof typeof colMapping;
const Cols = Object.keys(colMapping).map(value => Number(value)) as Col[];

interface Props {
    col?: number | null;
}

@Component({ name: componentName('TileGrid') })
@Api.Component('Tile Grid')
@Api.defaultSlot('Tiles displayed by the grid.')
export class TileGrid extends TsxComponent<Props> {
    @Api.Prop('number of columns', prop => prop.type(Number).acceptValues(...Cols))
    @Prop({ type: Number, default: null, required: false })
    public col!: Col | null;

    public render() {
        return <div class={this.classes}>{this.$slots.default}</div>;
    }

    private get classes() {
        const col = this.col;
        const colClass = col == null ? {} : { [`fd-tile-grid--${col}col`]: true };
        return {
            'fd-tile-grid': true,
            ...colClass,
        };
    }
}
