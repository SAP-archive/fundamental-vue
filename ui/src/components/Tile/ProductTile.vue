<template>
  <div class="fd-product-tile" :role="role" :aria-disabled="disabled">
    <div class="fd-product-tile__media" :style="style" />
    <div class="fd-product-tile__content">
      <h2 v-if="title != null" class="fd-product-tile__title">{{ title }}</h2>
      <slot name="content" />
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
export default Vue.extend({
  name: "FdProductTile",
  props: {
    url: { type: String, default: null },
    title: { type: String, default: null },
    isButton: { type: Boolean, default: null },
    disabled: { type: String, default: null }
  },
  computed: {
    style(): object {
      const url = this.url;
      if (url == null) {
        return {};
      }
      return {
        "background-image": `url(${this.url})`
      };
    },
    role(): string | null {
      return this.isButton ? "button" : null;
    }
  }
});
</script>
