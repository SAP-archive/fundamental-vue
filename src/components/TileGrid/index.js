import TileGrid from "./TileGrid.vue";
import { pluginify } from "./../../util";
export default pluginify(TileGrid);
export { TileGrid };
