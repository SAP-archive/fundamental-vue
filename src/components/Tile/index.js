import Tile from "./Tile.vue";
import ProductTile from "./ProductTile.vue";
import { pluginify } from "./../../util";
export default pluginify(Tile, ProductTile);
export { Tile, ProductTile };
