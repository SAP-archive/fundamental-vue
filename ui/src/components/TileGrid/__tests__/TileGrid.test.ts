import { shallowMount } from "@vue/test-utils";
import TileGrid from "../TileGrid.vue";
import Tile from "@/components/Tile";
import { CreateElement, VNode } from "vue";

describe("TileGrid", () => {
  describe("2 Columns", () => {
    const createTileElement = (h: CreateElement): VNode => {
      return h(Tile, {
        props: {
          description: "Lorem ipsum"
        }
      });
    };

    const wrapper = shallowMount(TileGrid, {
      propsData: {
        col: 2
      },
      slots: {
        default: {
          render(h: CreateElement): VNode {
            return h("div", {}, [
              createTileElement(h),
              createTileElement(h),
              createTileElement(h),
              createTileElement(h),
              createTileElement(h)
            ]);
          }
        }
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("has correct classes", () => {
      expect(wrapper.contains(".fd-tile-grid")).toBe(true);
    });

    it("renders as 2 column", () => {
      expect(wrapper.contains(".fd-tile-grid--2col")).toBe(true);
    });
  });

  describe("6 Columns", () => {
    const createTileElement = (h: CreateElement): VNode => {
      return h(Tile, {
        props: {
          description: "Lorem ipsum"
        }
      });
    };

    const wrapper = shallowMount(TileGrid, {
      propsData: {
        col: 6
      },
      slots: {
        default: {
          render(h: CreateElement): VNode {
            return h("div", {}, [
              createTileElement(h),
              createTileElement(h),
              createTileElement(h),
              createTileElement(h),
              createTileElement(h)
            ]);
          }
        }
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders as 6 column", () => {
      expect(wrapper.contains(".fd-tile-grid--6col")).toBe(true);
    });
  });
});
