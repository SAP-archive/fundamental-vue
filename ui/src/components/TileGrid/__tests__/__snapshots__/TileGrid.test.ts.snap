// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileGrid 2 Columns renders correctly 1`] = `
<div
  class="fd-tile-grid fd-tile-grid--2col"
>
  <div>
    <vuecomponent-stub
      description="Lorem ipsum"
    />
    <vuecomponent-stub
      description="Lorem ipsum"
    />
    <vuecomponent-stub
      description="Lorem ipsum"
    />
    <vuecomponent-stub
      description="Lorem ipsum"
    />
    <vuecomponent-stub
      description="Lorem ipsum"
    />
  </div>
</div>
`;

exports[`TileGrid 6 Columns renders correctly 1`] = `
<div
  class="fd-tile-grid fd-tile-grid--6col"
>
  <div>
    <vuecomponent-stub
      description="Lorem ipsum"
    />
    <vuecomponent-stub
      description="Lorem ipsum"
    />
    <vuecomponent-stub
      description="Lorem ipsum"
    />
    <vuecomponent-stub
      description="Lorem ipsum"
    />
    <vuecomponent-stub
      description="Lorem ipsum"
    />
  </div>
</div>
`;
